### PR TITLE
CXX XO example and fixed CTRL-C handling

### PR DESCRIPTION
**Pull request by EugeneYYY**
Added CXX XO example
Fixed CRTL-C handler and TP unregistration on exit
Modified AdrresMapper API to make it more generic

### CODE REVIEW

File: sdk/examples/intkey_cxx/address_mapper.cpp [hunk 1]
```diff
@@ -25,9 +25,7 @@
 #include "cryptopp/hex.h"
 #include "log4cxx/logger.h"
 
-#include "proto/state_context.pb.h"
-
-namespace sawtooth {
+//#include "proto/state_context.pb.h"
```
Comment (vaporos) [inline sdk/examples/intkey_cxx/address_mapper.cpp]: Remove commented-out code.

File: sdk/examples/intkey_cxx/address_mapper.h [hunk 2]
```diff
@@ -44,18 +43,19 @@ class AddressFormatError: public std::runtime_error {
 // namespace prefix when it is constructed. If your transaction processor
 // operates on multiple namespaces, it is recommended that you use an instance
 // for each namespace.
-class AddressMapperImpl: public AddressMapper {
+class AddressMapper {
  public:
     // Constructor for an address mapping object, Takes the
     // unencoded namespace name, it will be mapping objects, to as a parameter.
     // The namespace argument passed will be mapped to a namespace_prefix
+    //explicit AddressMapper(const std::string& namespace_);
```
Comment (vaporos) [inline sdk/examples/intkey_cxx/address_mapper.h]: Remove commented-out code.

File: sdk/cxx/src/transaction_processor.cpp [hunk 3]
```diff
@@ -33,6 +33,34 @@ namespace sawtooth {
...
+static bool s_interrupt_initialized = false;
+
+static void s_signal_handler (int signal_value) {
+    s_interrupted = 1;
+    LOG4CXX_ERROR(logger, "signal interrupt received");
+}
+
+static void s_catch_signals (void) {
+    if (!s_interrupt_initialized)
+    {
```
Comment (vaporos) [inline sdk/cxx/src/transaction_processor.cpp]: The curly brace placement is not consistent with the rest of the code.

File: sdk/cxx/src/transaction_processor.cpp [hunk 4]
```diff
@@ -33,6 +33,32 @@ namespace sawtooth {
...
+        action.sa_handler = s_signal_handler;
+        action.sa_flags = 0;
+        sigemptyset (&action.sa_mask);
+        sigaction (SIGINT, &action, NULL);
+        sigaction (SIGTERM, &action, NULL);
+        s_interrupt_initialized = true;
+
+        LOG4CXX_ERROR(logger, "signal handler initialized");
+    }
+    else {
```
Comment (cianx) [inline sdk/cxx/src/transaction_processor.cpp]: Curly braces inconsistant with the rest of the code.

File: sdk/cxx/src/transaction_processor.cpp [hunk 5]
```diff
@@ -79,6 +105,26 @@ void TransactionProcessorImpl::Register() {
...
+
+    TpUnregisterResponse response;
+    future->GetMessage(Message_MessageType_TP_UNREGISTER_RESPONSE, &response);
+
+    if (response.status() != TpUnregisterResponse::OK) {
+        LOG4CXX_ERROR(logger, "Unregister failed, status code: "
+                    << response.status());
+    }
+    else
+    {
```
Comment (cianx) [inline sdk/cxx/src/transaction_processor.cpp]: Curly braces inconsistant with the rest of the code.

File: sdk/examples/intkey_cxx/address_mapper.h [hunk 6]
```diff
@@ -17,17 +17,16 @@
 #include <memory>
 #include <string>
 
-#include "sawtooth_sdk.h"
-#include "sawtooth/message_stream.h"
-
-namespace sawtooth {
+//#include "sawtooth_sdk.h"
+//#include "sawtooth/message_stream.h"
```
Comment (cianx) [inline sdk/examples/intkey_cxx/address_mapper.h]: Please remove vommented code.

File: sdk/examples/xo_cxx/address_mapper.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <memory>
+#include <string>
+
+//#include "sawtooth_sdk.h"
+//#include "sawtooth/message_stream.h"
```
Comment (cianx) [inline sdk/examples/xo_cxx/address_mapper.h]: Please remove commented code.

File: sdk/examples/xo_cxx/address_mapper.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+ public:
+    // Constructor for an address mapping object, Takes the
+    // unencoded namespace name, it will be mapping objects, to as a parameter.
+    // The namespace argument passed will be mapped to a namespace_prefix
+    AddressMapper(const std::string& namespace_);
+    virtual ~AddressMapper() {}
+
+    //AddressMapper (const AddressMapper&) = delete;
+    //AddressMapper (const AddressMapper&&) = delete;
+    //AddressMapper& operator= (const AddressMapper&) = delete;
```
Comment (cianx) [inline sdk/examples/xo_cxx/address_mapper.h]: Please remove commented code.

File: sdk/examples/xo_cxx/xo_transaction_processor.cpp [hunk 9]
```diff
@@ -0,0 +1,705 @@
...
+        }
+        catch (...) {
+            throw InvalidTransaction(
+                "Invalid Action: Failed to to make address");
+        }
+    };
+
+    // load current state for the game name
+    bool LoadState(const string& name) {
+        bool bOk = false;
```
Comment (cianx) [inline sdk/examples/xo_cxx/xo_transaction_processor.cpp]: the name "ok" is preferable.

File: sdk/examples/xo_cxx/xo_transaction_processor.cpp [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+void ParseArgs(int argc, char** argv, string& connectStr) {
+    bool bLogLevelSet = false;
+
+    for (int i = 1; i < argc; i++) {
+        const char* arg = argv[i];
+        if (!strcmp(arg, "-h") || !strcmp(arg, "--help")) {
+            Usage(true, 0);
+        }
+        else
+        if (!strcmp(arg, "-v")) {
```
Comment (cianx) [inline sdk/examples/xo_cxx/xo_transaction_processor.cpp]: Curly braces inconsistant with the rest of the code.

**Pull request merged**